### PR TITLE
GFORMS-2110: Add support to allow a dividerPosition in a dynamic choi…

### PR DESCRIPTION
**Pull request by lxol**
…ce component

### CODE REVIEW

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/ComponentType.scala [hunk 1]
```diff
@@ -260,7 +260,7 @@ case class Choice(
   selections: List[Int],
   hints: Option[NonEmptyList[SmartString]],
   optionHelpText: Option[NonEmptyList[SmartString]],
-  dividerPosition: Option[Int],
+  dividerPosition: Option[String],
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/sharedmodel/formtemplate/ComponentType.scala:263]: Please introduce proper type (ADT) for this.

**Review by josef-vlach — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/ComponentType.scala (hunk 1 above)
Comment (lxol) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/ComponentType.scala:263]: I thought ADT is a bit overkill for this case.

File: app/uk/gov/hmrc/gform/sharedmodel/formtemplate/ComponentType.scala [hunk 2]
```diff
@@ -253,14 +253,22 @@ object NoneChoice {
   implicit val format: OFormat[NoneChoice] = OFormatWithTemplateReadFallback(templateReads)
 }
 
+sealed trait DividerPosition
+final case class IntDivider(pos: Int) extends DividerPosition
+final case class StringDivider(pos: String) extends DividerPosition
```
[the remark this answers is not in the document]
Comment (lxol) [reply app/uk/gov/hmrc/gform/sharedmodel/formtemplate/ComponentType.scala]: done

**Review by josef-vlach — CHANGES_REQUESTED** [review]
Missing validation for `DividerPosition.Value`. Any value is now accepted.

**Comment by lxol** [comment]
> Missing validation for `DividerPosition.Value`. Any value is now accepted.

done

File: app/uk/gov/hmrc/gform/formtemplate/FormComponentMaker.scala [hunk 3]
```diff
@@ -157,7 +157,11 @@ class FormComponentMaker(json: JsValue) {
       case JsUndefined() => Right(None)
     }
 
-  lazy val dividerPositon: Option[Int] = (json \ "dividerPosition").asOpt[Int]
+  lazy val dividerPositon: Option[DividerPosition] =
+    for {
+      s <- (json \ "dividerPosition").asOpt[String]
+    } yield Try(s.toInt).toOption.fold[DividerPosition](DividerPosition.Value(s))(i => DividerPosition.Number(i))
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/formtemplate/FormComponentMaker.scala:163]: Your comment [here](https://github.com/hmrc/gform/pull/918#discussion_r1142011679) pique my interest. 
Int and String are actually easily distinguishable, and we should not do `Try(s.toInt)`

Mapping:
```
"dividerPosition": 4, // -> DividerPosition.Number
"dividerPosition": "4", // -> DividerPosition.Value
```
Given this mapping we can match on json types without need for `Try`

**Review by josef-vlach — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/gform/formtemplate/FormComponentMaker.scala (hunk 3 above)
Comment (lxol) [reply app/uk/gov/hmrc/gform/formtemplate/FormComponentMaker.scala]: I refactored to remove Try

**Pull request merged**